### DISCUSSION TITLE
Warnings: asyncData should return an object RefImpl

### DISCUSSION THREAD

**Discussion by vedmant** [nuxt/nuxt#16012] [Questions]
I have warnings in the console like:

```
[nuxt] asyncData should return an object RefImpl {                                                                                                                            16:55:29
  __v_isShallow: false,
  dep: undefined,
  __v_isRef: true,
  _rawValue: null,
  _value: null
}
```

What I'm doing wrong? Here is my code:

```vue
<template>
  <NuxtPage />
</template>

<script lang="ts">
import { defineNuxtComponent, useRoute } from '#imports'
import { useOrderStore } from '~/stores/order'

export default defineNuxtComponent({
  async asyncData ({ $pinia }) {
    const route = useRoute()
    const store = useOrderStore($pinia)

    const uuid = route.params.orderUuid
    await store.getOrder(uuid)

    return {}
  },
})
</script>
```

**Comment by IsraelOrtuno** [comment]
I didn't even know you could use that `asyncData` in components as most of the examples in the current documentation are using `<script setup>` way. 

Have you tried it?

```vue
<script setup lang="ts">
const route = useRoute();
const store = useOrderStore();
const uuid = route.params.orderUuid

const { data: order } = await useAsyncData(() => store.getOrder(uuid));
</script>
```

**Reply by vedmant** [reply]
@IsraelOrtuno This is composition API, I'm using options API which Vue and Nuxt also supports, but Nuxt has basically no documentation how to use options API

**Reply by IsraelOrtuno** [reply]
I am aware of that but since Nuxt itself recommends using Composition API or Script setup (https://nuxt.com/docs/migration/overview#overview) I would recommend to switch to that since I do not think there will be efforts for documenting that anytime soon. Your code is almost 100% Composition API as you are already using composables.

**Reply by vedmant** [reply]
@IsraelOrtuno I'm not going to switch to composition API because someone recommends it, when Vue 3 introduced composition API it was said clearly that options API will not be going to deprecate so this is absolutely normal way of coding with Vue 3. And lack of documentation for options API in Nuxt 3 is not a reason to switch to composition API. 

I still believe composition API is a terrible idea for smaller projects, will cause lots of code spaghetti and reduce code readability, especially when mid class developer will work on it, instead of thinking about proper components architecture, people will start writing large frankenstein components mixing everything together. Options API allowed clean and strict code structure that's enforced to follow and enforced to split components to smaller with better design in general.

My code is still options API, I'm using some of use* cause there is no other way to do this in Nuxt as far as I can tell.

**Reply by meirroth** [reply]
I second the above.
We've encountered the same issue while migrating our app. We'd like to stick with Options API and wish Nuxt will add support and docs.

**Comment by harish2704** [comment]
Hi all,

I was also facing this same issue while using Nuxt 3.0.0. 

After a bit of debugging, I found that issue is caused due [this line](https://github.com/nuxt/framework/blob/9c478ee16eea288960a4a9a07caca941507c1309/packages/nuxt/src/app/composables/asyncData.ts#L113) . 

What happens is that, actually `getCachedData()` function will return `null` if it is not cached but we are comparing it against `undefined` .  Thus `hasCachedData()` will evaluate to wrong value. 

When it happens application will not immediately execute or call component's `asyncData` function . 

it got fixed when I changed the line to 
```javascript
const hasCachedData = () => getCachedData() != null;
```

( **Please note**:  I don't know the internals of Nuxt3 . So my explanation can be wrong. but it fixed the issue for me )

**Comment by memic84** [comment]
Do you have a reproduction?

Have you tried something like this? Do you get the same result?
```vue
<template>
  <NuxtPage />
</template>

<script lang="ts">
import { defineNuxtComponent, useRoute } from '#imports'
import { useOrderStore } from '~/stores/order'

export default defineNuxtComponent({
  data() {
	return {
		order: []
	}
  },
  async fetch ({ $pinia }) {
    const route = useRoute()
    const store = useOrderStore($pinia)

    const uuid = route.params.orderUuid

    this.order = await store.getOrder(uuid)
  },
})
</script>
```

**Comment by harish2704** [comment]
My previous message was wrong. 

For me, this issue happened due incorrectly configured `SSR` functionality. 
The issue is , when `SSR` is enabled (which is the default behavior for a new next3 app ),Nuxt will execute `asyncData` function from server side . 
If we didn't address this case by properly forwarding cookie / token then the request sent from server will fail . That was the cause in the case of mine.

I found this while tried to create a small project for reproducing this issue.

In short, this issue will disappear if simply disable ssr in nuxt config
```javascript
export default defineNuxtConfig({
  ssr:  false,
});
```

**Reply by vedmant** [reply]
The whole point of `asyncData()` is to execute it on backend for SSR. If you don't need SSR, you just can run everything you need in `mounted()`

**Reply by harish2704** [reply]
Thanks for the reply. I got that point. 
For me, I had a bigger problem. My backend API was writen in ExpressJS and there was a bug while calling Express API routes from Nuxt3's SSR .
I found this while digging deep into the issue. 
it's details can be found here https://github.com/unjs/nitro/issues/351#issuecomment-1345575954 

So, when this bug is fixed, I can SSR  again.

**Comment by KartoffelCheetah** [comment]
This error still occurs on the latest version of nuxt when an asyncData method written in OptionsAPI is defined on a component but the fetchKey on the client side is not matching on the fetchKey on the server side.

TLDR;

**You have to define the OptionsAPI fetchKey method or propertyin a way that it doesn't change between the browser and the client**. It can be a "_string_" of your choice or any "_function returning a string_". It can be even as simple as 'Hello' or () => 'Hello' if you don't mind multiple components reusing the same data.

Here is a more complex solution which does the same as nuxt's default implementation, but fixes a nuxt "bug" where the usage of url fragments always cause a _cache miss_:

```js
// myComponent.js
import { fetchKey } from './fetchKeyFixForUrlFragment.js';
export default {
  fetchKey,
  asyncData () { /* ... */ },
  /* ... */
};
```
```js
// fetchKeyFixForUrlFragment.js
import { getCurrentInstance } from 'vue';
import { useRoute } from 'nuxt/app';

export function fetchKey () {
  // Custom implementation of node_modules/nuxt/dist/app/composables/component.js:8 runLegacyAsyncData
  // to solve this issue: https://github.com/nuxt/nuxt/discussions/16012#discussion-4773317
  // NuxtJS includes the route.fullPath when fetchKey is not defined,
  // but route.fullPath contains the url fragment only on the client side so
  // that causes a cache miss.
  const instance = getCurrentInstance();
  const route = useRoute();
  const generatedKey = [instance.type._fetchKeyBase, route.fullPath.replace(route.hash, ''), route.matched.findIndex(r => Object.values(r.components || {}).includes(instance.type))].join(':');
  return generatedKey;
}

```

(sorry, I'm very bad at TLDRs)

---

## Where does the warning comes from

The warning is logged by this line: `console.warn('[nuxt] asyncData should return an object', data)`.

inside the `runLegacyAsyncData` function, which is the function that nuxt3 uses to handle the OptionsAPI style asyncData

https://github.com/nuxt/nuxt/blob/3c271413f85b1dcc8bf41714512627ccebbb6c75/packages/nuxt/src/app/composables/component.ts#L12

runLegacyAsyncData internally uses the newer CompositionAPI function `useAsyncData` https://github.com/nuxt/nuxt/blob/3c271413f85b1dcc8bf41714512627ccebbb6c75/packages/nuxt/src/app/composables/asyncData.ts#L190

The problem here is that useAsyncData method requires a fetchKey as a first parameter to correlate the data on the server side and client side. If the fetchKey on the client side does not match the one on the SSR side then the `getDefaultCachedData` function and the whole useAsyncData would return null resulting in the said warning.

**To fix this you have to make sure that key matches on both sides!** This is the current implementation in runLegacyAsyncData:

```ts
const key = (
  (typeof fetchKey === 'function' ? fetchKey(() => '') : fetchKey) ||
  ([_fetchKeyBase, route.fullPath, route.matched.findIndex(r => Object.values(r.components || {}).includes(vm.type))].join(':'))
);
```

This is sadly a buggy implementation so there is always a chance for it to be mismatching. A very simple example would be when the url contains an urlFragment (or hash). Since browsers never send the fragment to the server, in this case the route.fullPath would be always different.
For example opening `https://example.com` would work fine, but opening `https://example.com#x` would show this warning.

The only real way to fix this warning is to specify fetchKey explicitely yourself on your components beside asyncData. See one possible implementation above.

**Reply by nseam** [reply]
I spent 4 hours trying to make `asyncData()` to work using Options API (along with `nuxt3-class-component` package that allows me to use decorators). Adding `fetchKey() { return '<something>' }` fixed the problem with `asyncData should return an object` error. Thank you very much!

**Reply by KartoffelCheetah** [reply]
I'm glad it helped! :+1:
